### PR TITLE
Update C++ standard to C++17

### PR DESCRIPTION
**Pull request by kparzysz-quic**
LLVM has switched to C++17 in its development branch. Follow suit to be able to compile LLVM headers.

### CODE REVIEW

**Comment by tqchen** [comment]
LGTM! @kparzysz-quic do you mind also post an announcement thread? So people are aware

**Review by junrushao1994 — APPROVED** [review]
Ah so finally we are using C++17 in TVM 🎉

**Comment by kparzysz-quic** [comment]
> LGTM! @kparzysz-quic do you mind also post an announcement thread? So people are aware

Sure, I'll just need to resolve the build failures.  I'll post the announcement once the PR is ready to merge.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -36,14 +36,16 @@ concurrency:
 
 jobs:
   MacOS:
-    if: ${{ github.repository == 'apache/tvm' }}
+    if: ${{ github.repository == 'kparzysz-quic/tvm' }}
```
Comment (shingjan) [inline .github/workflows/main.yml:39]: Should we keep this change local? Otherwise LGTM!

Comment (kparzysz-quic) [reply .github/workflows/main.yml:39]: No, I'm trying to figure out the macOS build failure.  I'll clean this PR up before merging.

**Comment by ganler** [comment]
Great news!

Just need to carefully update the installation documents as well. For example, update "C++ 14" to "C++ 17" in https://github.com/apache/tvm/blob/fbb7b5d1a0d82acb1f581dd2ec362b4dcad2638e/docs/install/from_source.rst

In the meantime, it might be better to specify the required minimal compiler [versions](https://en.cppreference.com/w/cpp/compiler_support/17) as the coverage to C++17 standard can be quite different in different compiler versions.

For example, if we just want to pass the "-std=c++17" flag, then the required minimals can be:

- gcc >= 5
- clang >= 5
- visual studio >= 2017 15.3

However, those initial version might badly support some popular library features like `std::filesystem`. For example, gcc implements filesystems as a non-experimental library in [gcc-8](https://gcc.gnu.org/gcc-8/changes.html).

We may reference how other open-source projects moves to C++17. For example, [LLVM](https://discourse.llvm.org/t/important-new-toolchain-requirements-to-build-llvm-will-most-likely-be-landing-within-a-week-prepare-your-buildbots/61447) enables C++17 on Aprl 8 this year with several compiler version [constraints](https://github.com/llvm/llvm-project/commit/4c72deb613d9d8838785b431facb3eb480fb2f51).

cc: @junrushao1994 @tqchen

**Review by echuraev — APPROVED** [review]
Happy to see C++17 in TVM :tada:

**Review by shingjan — APPROVED** [review]
LGTM. Thanks for sending the PR!

**Comment by junrushao1994** [comment]
I would love to second @ganler that we might want to provide clearer instruction on TVM official website to make TVM more approachable to new contributors. @kparzysz-quic @tqchen would either of you mind making such a change accordingly? I'm happy to submit a PR too if you guys prefer

**Comment by kparzysz-quic** [comment]
I created a https://github.com/apache/tvm/pull/12405 with docs update.  Let me know if you want to have more information in it.

**Pull request merged**